clang/lib/AST/FormatString.cpp

```diff
@@ -484,6 +484,36 @@ ArgType::matchesType(ASTContext &C, QualType argTy) const {
         return NoMatch;
       }
 
+    case CCapabilityTy:
+      if (const PointerType *PT = argTy->getAs<PointerType>()) {
+        if (!PT->isCHERICapability()) {
+          // XXX: CheriBSD manually passes capabilities indirectly for hybrid
+          // printf, so permit pointers to capabilities for now, especially
+          // until the Morello toolchain catches up.
+          if (!C.getTargetInfo().areAllPointersCapabilities()) {
+            const QualType PointeeTy = PT->getPointeeType();
+            const PointerType *PointeePT = PointeeTy->getAs<PointerType>();
+            if (!PointeePT || !PointeePT->isCHERICapability())
+              return NoMatch;
```

Comment:
```suggestion
            if (!PointeeTy->isCapabilityPointerType())
              return NoMatch;
```

Reply:
That would also allow atomic capabilities.

Reply:
Yes, but is that a problem? It seems neither clang nor GCC warn for that case: https://godbolt.org/z/Ko9o6c